gflownet/gflownet.py

```diff
@@ -524,68 +532,61 @@ def flowmatch_loss(self, it, batch, loginf=1000):
             Loss of the intermediate nodes only
         """
         loginf = self._tfloat([loginf])
-        batch_idxs = self._tlong(
-            sum(
-                [
-                    [i] * len(parents)
-                    for i, (_, _, _, parents, _, _, _, _, _) in enumerate(batch)
-                ],
-                [],
-            )
+        # Unpack batch
+        (
+            states,
+            actions,
+            parents,
+            parents_a,
+            done,
+            traj_id_parents,
+            state_id,
+            masks_sf,
+            masks_b,
+        ) = zip(*batch)
+        # Get state/batch id
+        parents_batch_id = self._tlong(
+            sum([[idx] * len(p) for idx, p in enumerate(parents)], [])
         )
-        sp, _, r, parents, actions, done, _, _, masks = map(torch.cat, zip(*batch))
-        # Sanity check if negative rewards
-        if self.logger.debug and torch.any(r < 0):
-            neg_r_idx = torch.where(r < 0)[0].tolist()
-            for idx in neg_r_idx:
-                state_oracle = self.env.state2oracle(sp)
-                output_proxy = self.env.proxy(state_oracle)
-                reward = self.env.proxy2reward(output_proxy)
-                import ipdb
-
-                ipdb.set_trace()
-
-        # Q(s,a)
-        parents_Qsa = self.forward_policy(parents)[
-            torch.arange(parents.shape[0]), actions
-        ]
-
-        # log(eps + exp(log(Q(s,a)))) : qsa
-        in_flow = torch.log(
-            self.loss_eps
-            + self._tfloat(torch.zeros((sp.shape[0],))).index_add_(
-                0, batch_idxs, torch.exp(parents_Qsa)
-            )
+        # Concatenate lists of tensors
+        states, parents, parents_a, done, masks_sf = map(
```

Comment:
I think we might have to deal with `states` and `parents` separately for variable length sequences as we did for `tb`

Reply:
uhm but that actually broke a bunch of things :sweat_smile: let's deal with it when the need arises :grimacing:

Reply:
Yeah so in [PR#68](https://github.com/alexhernandezgarcia/gflownet/pull/68/files#diff-700050cd4007666a2c357edc47c03336bdbaddab3e135d60a3f382718580780b), I slightly modify that implementation by introducing a `do_padding` variable in the env, which is `False` by default (so that it does not break the grid/torus envs) and can be explicitly set to `True` for AMP

Reply:
Ah that sounds better. I think we should merge this for now and you can add that part to the flowmatch loss if you ever need to use it?